### ISSUE TITLE
feat: make FFI work in the arm64 simulator on arm64

### ISSUE THREAD

**Issue by eseidel** [shorebirdtech/shorebird#532]
Just using this bug to record notes.

https://github.com/dart-lang/sdk/issues/37299 is the dart bug.

```
eseidel@erics-mbp sdk % git diff HEAD                                        
diff --git a/runtime/lib/ffi_dynamic_library.cc b/runtime/lib/ffi_dynamic_library.cc
index 489652142fc..21d82581437 100644
--- a/runtime/lib/ffi_dynamic_library.cc
+++ b/runtime/lib/ffi_dynamic_library.cc
@@ -29,7 +29,7 @@
 
 namespace dart {
 
-#if defined(USING_SIMULATOR) || (defined(DART_PRECOMPILER) && !defined(TESTING))
+#if (defined(DART_PRECOMPILER) && !defined(TESTING))
 
 DART_NORETURN static void SimulatorUnsupported() {
 #if defined(USING_SIMULATOR)
diff --git a/runtime/platform/globals.h b/runtime/platform/globals.h
index 40d8ac3d1cf..d11ee063f7c 100644
--- a/runtime/platform/globals.h
+++ b/runtime/platform/globals.h
@@ -381,6 +381,8 @@ struct simd128_value_t {
 #error Unknown architecture.
 #endif
 
+#define USING_SIMULATOR 1
+
 #if !defined(DART_TARGET_OS_ANDROID) && !defined(DART_TARGET_OS_FUCHSIA) &&    \
     !defined(DART_TARGET_OS_MACOS_IOS) && !defined(DART_TARGET_OS_LINUX) &&    \
     !defined(DART_TARGET_OS_MACOS) && !defined(DART_TARGET_OS_WINDOWS)

```

```
Process 62606 launched: '/Users/eseidel/Documents/GitHub/dart-sdk/sdk/xcodebuild/DebugARM64/dart_precompiled_runtime_product' (arm64)
Simulator hit Unimplemented instruction: at 0x101cbbf88, last_pc=0x101cbbf84

../../runtime/vm/handles_impl.h: 84: error: expected: thread->MayAllocateHandles()
version=3.1.0-edge.86c18251b462a68b64a00e14fb39e292441193ff (be) (Mon May 22 15:17:07 2023 +0000) on "macos_simarm64"
pid=62606, thread=11267, isolate_group=main(0x104808200), isolate=main(0x104810e00)
os=macos, arch=arm64, comp=no, sim=yes
isolate_instructions=101be6800, vm_instructions=101be0000
fp=170559c80, sp=170559c58, pc=1003500c8
  pc 0x00000001003500c8 fp 0x0000000170559c80 dart::Profiler::DumpStackTrace(void*)+0x68
  pc 0x00000001000575a0 fp 0x0000000170559ca0 dart::Assert::Fail(char const*, ...) const+0x28
  pc 0x000000010021e08c fp 0x0000000170559ce0 dart::VMHandles::AllocateZoneHandle(dart::Zone*)+0x0
  pc 0x0000000100167d28 fp 0x0000000170559d20 dart::Object::HandleImpl(dart::Zone*, dart::ObjectPtr, long)+0x1c
  pc 0x00000001003c1688 fp 0x000000017055a090 dart::SimulatorDebugger::Debug()+0x178
  pc 0x00000001003c509c fp 0x000000017055a160 dart::Simulator::DecodePCRel(dart::Instr*)+0xe0
  pc 0x00000001003c32b8 fp 0x000000017055a240 dart::Simulator::InstructionDecode(dart::Instr*)+0x140
  pc 0x00000001003cab6c fp 0x000000017055a340 dart::Simulator::Execute()+0x80
  pc 0x00000001003caeb8 fp 0x000000017055a430 dart::Simulator::Call(long long, long long, long long, long long, long long, bool, bool)+0x254
  pc 0x00000001002005b4 fp 0x000000017055a4c0 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&)+0x108
  pc 0x0000000100203378 fp 0x000000017055a530 dart::DartLibraryCalls::HandleMessage(long long, dart::Instance const&)+0x10c
  pc 0x0000000100227f04 fp 0x000000017055acf0 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message>>)+0x33c
  pc 0x000000010023a98c fp 0x000000017055adb0 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x1ec
  pc 0x000000010023b65c fp 0x000000017055ae40 dart::MessageHandler::TaskCallback()+0x2dc
  pc 0x00000001003e8564 fp 0x000000017055aef0 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x17c
  pc 0x00000001003e8e24 fp 0x000000017055af50 dart::ThreadPool::Worker::Main(unsigned long)+0x124
  pc 0x000000010034b568 fp 0x000000017055afc0 dart::ThreadStart(void*)+0xcc
  pc 0x00000001af91a06c fp 0x000000017055afe0 _pthread_start+0x94
-- End of DumpStackTrace
../../runtime/vm/stack_frame.cc: 357: error: expected: code != Code::null()
Aborting reentrant request for stack trace.
Process 62606 stopped
* thread #9, name = 'DartWorker', stop reason = signal SIGABRT
    frame #0: 0x00000001af8e2868 libsystem_kernel.dylib`__pthread_kill + 8
libsystem_kernel.dylib`:
->  0x1af8e2868 <+8>:  b.lo   0x1af8e2888               ; <+40>
    0x1af8e286c <+12>: pacibsp 
    0x1af8e2870 <+16>: stp    x29, x30, [sp, #-0x10]!
    0x1af8e2874 <+20>: mov    x29, sp
Target 0: (dart_precompiled_runtime_product) stopped.
```

**Comment by eseidel** [comment]
Release:
```eseidel@erics-mbp ffi_test % ../dart-sdk/sdk/xcodebuild/ReleaseARM64/dart-sdk/bin/dart compile aot-snapshot ffi.dart
Generated: /Users/eseidel/Documents/GitHub/ffi_test/ffi.aot
eseidel@erics-mbp ffi_test % lldb ../dart-sdk/sdk/xcodebuild/ReleaseARM64/dart_precompiled_runtime_product ffi.aot
(lldb) target create "../dart-sdk/sdk/xcodebuild/ReleaseARM64/dart_precompiled_runtime_product"
Current executable set to '/Users/eseidel/Documents/GitHub/dart-sdk/sdk/xcodebuild/ReleaseARM64/dart_precompiled_runtime_product' (arm64).
(lldb) settings set -- target.run-args  "ffi.aot"
(lldb) r
Process 71315 launched: '/Users/eseidel/Documents/GitHub/dart-sdk/sdk/xcodebuild/ReleaseARM64/dart_precompiled_runtime_product' (arm64)
Simulator hit Unimplemented instruction: at 0x1005cbf88, last_pc=0x1005cbf84

Disassembler not supported in this mode.
sim> 
```

Note this is all on dart-lang/sdk@86c18251b462a68b64a00e14fb39e292441193ff

**Comment by eseidel** [comment]
I made it work.  Code would need a lot more work to be landable (including supporting cross-arch calls), but is enough to unblock the next phase of my investigation: https://github.com/shorebirdtech/dart-sdk/tree/simulator_ffi. Thanks for the help!

**Comment by eseidel** [comment]
Had to move to a newer Dart (since the 3.0.1 branch doesn't build cleanly from the Flutter engine?)
https://github.com/flutter/flutter/issues/127606

Now hitting a different FFI bug, investigating:

```
[        ] 2023-05-25 15:13:24.440885-0700 Runner[36107:3774704] [VERBOSE0:FlutterViewController.mm(235)] CPU::Id(): simarm64
[  +71 ms] 2023-05-25 15:13:24.513155-0700 Runner[36107:3774704] [VERBOSE-2:FlutterDarwinContextMetalImpeller.mm(39)] Using the Impeller rendering backend.
[ +119 ms] 2023-05-25 15:13:24.631709-0700 Runner[36107:3774740] [VERBOSE-2:dart_vm_initializer.cc(41)] Unhandled Exception: type 'int' is not a subtype of type 'bool' of 'boolean expression'
[        ] #0      PlatformDispatcher.implicitView (dart:ui/platform_dispatcher.dart:207:36)
[        ] #1      RendererBinding.createViewConfiguration (package:flutter/src/rendering/binding.dart:303:49)
[        ] #2      RendererBinding.initRenderView (package:flutter/src/rendering/binding.dart:215:44)
[        ] #3      RendererBinding.initInstances (package:flutter/src/rendering/binding.dart:41:5)
[        ] #4      WidgetsBinding.initInstances (package:flutter/src/widgets/binding.dart:278:11)
[        ] #5      new BindingBase (package:flutter/src/foundation/binding.dart:151:5)
[        ] #6      new _WidgetsFlutterBinding&BindingBase&GestureBinding (package:flutter/src/widgets/binding.dart)
[        ] #7      new _WidgetsFlutterBinding&BindingBase&GestureBinding&SchedulerBinding (package:flutter/src/widgets/binding.dart)
[        ] #8      new _WidgetsFlutterBinding&BindingBase&GestureBinding&SchedulerBinding&ServicesBinding (package:flutter/src/widgets/binding.dart)
[        ] #9      new _WidgetsFlutterBinding&BindingBase&GestureBinding&SchedulerBinding&ServicesBinding&PaintingBinding (package:flutter/src/widgets/binding.dart)
[        ] #10     new _WidgetsFlutterBinding&BindingBase&GestureBinding&SchedulerBinding&ServicesBinding&PaintingBinding&SemanticsBinding (package:flutter/src/widgets/binding.dart)
[        ] #11     new _WidgetsFlutterBinding&BindingBase&GestureBinding&SchedulerBinding&ServicesBinding&PaintingBinding&SemanticsBinding&RendererBinding (package:flutter/src/widgets/binding.dart)
[        ] #12     new _WidgetsFlutterBinding&BindingBase&GestureBinding&SchedulerBinding&ServicesBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding (package:flutter/src/widgets/binding.dart)
[        ] #13     new WidgetsFlutterBinding (package:flutter/src/widgets/binding.dart)
[        ] #14     WidgetsFlutterBinding.ensureInitialized (package:flutter/src/widgets/binding.dart:1319:7)
[        ] #15     runApp (package:flutter/src/widgets/binding.dart:1092:56)
[        ] #16     main (package:ios_test/main.dart:5:3)
[        ] #17     _runMain.<anonymous closure> (dart:ui/hooks.dart:159:23)
[        ] #18     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:296:19)
[        ] #19     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:189:12)
```

**Comment by eseidel** [comment]
I'm currently using these versions and commands:
```

Flutter is already up to date on channel main
Flutter 3.11.0-14.0.pre.15 • channel main • https://github.com/flutter/flutter.git
Framework • revision 8576da53c5 (9 minutes ago) • 2023-05-25 14:01:18 -0700
Engine • revision 6ff02c1726
Tools • Dart 3.1.0 (build 3.1.0-146.0.dev) • DevTools 2.23.1



./flutter/tools/gn --no-goma --unoptimized --no-prebuilt-dart-sdk
ninja -C out/host_debug_unopt 


./flutter/tools/gn --no-goma --unoptimized --no-prebuilt-dart-sdk --ios --mac-cpu=arm64 --gn-arg='dart_force_simulator=true'
ninja -C out/ios_debug_unopt_arm64
```

**Comment by eseidel** [comment]
https://github.com/flutter/flutter/issues/127606 was another issue I had to work around (by using xcode):
`flutter build ios --config-only --local-engine-src-path=$HOME/Documents/GitHub/ios_engine/src --local-engine ios_debug_unopt_arm64`

**Comment by eseidel** [comment]
Next step is to try and repro the failure seen at:
https://github.com/flutter/engine/blob/main/lib/ui/platform_dispatcher.dart#L209

**Comment by eseidel** [comment]
I have succeeded. :)

```
2023-05-26 12:58:23.622585-0700 Runner[40902:4302177] [VERBOSE0:FlutterViewController.mm(235)] CPU::Id(): simarm64
2023-05-26 12:58:23.622761-0700 Runner[40902:4302177] Metal API Validation Enabled
2023-05-26 12:58:23.741828-0700 Runner[40902:4302177] [VERBOSE-2:FlutterDarwinContextMetalImpeller.mm(39)] Using the Impeller rendering backend.
2023-05-26 12:58:46.809104-0700 Runner[40902:4302551] flutter: The Dart VM service is listening on http://127.0.0.1:49356/tnKqhWhHIdk=/
2023-05-26 12:58:46.937979-0700 Runner[40902:4302530] flutter: Log from dart
2023-05-26 12:58:56.034026-0700 Runner[40902:4302530] flutter: incrementing counter
2023-05-26 12:58:56.216961-0700 Runner[40902:4302530] flutter: incrementing counter
2023-05-26 12:58:56.375327-0700 Runner[40902:4302530] flutter: incrementing counter
2023-05-26 12:58:56.525322-0700 Runner[40902:4302530] flutter: incrementing counter
2023-05-26 12:58:56.657601-0700 Runner[40902:4302530] flutter: incrementing counter
2023-05-26 12:58:56.815126-0700 Runner[40902:4302530] flutter: incrementing counter
```

**Comment by eseidel** [comment]
My bug was that I was including the .c stub trampoline on arm64, which was just failing to do anything: https://github.com/shorebirdtech/dart-sdk/commit/6dc6976b17e85417f3fda57377d9418bd631f473

**Comment by yaminet1024** [comment]
@eseidel hi, will this part of the code be shared? very interested.

**Comment by eseidel** [comment]
> @eseidel hi, will this part of the code be shared? very interested.

I would like to eventually open source / upstream it all, but probably not in 2024.
